text/0000-alpha-and-omega.md

```diff
@@ -0,0 +1,109 @@
+- Feature Name: Higher and Lower Types
+- Start Date: 2017-04-01 
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+[summary]: #summary
+
+This proposal enables programmers to continuously navigate the performance-expressiveness curve by providing a mechanism for explicitly working with higher and lower types ("Omega Types").
+
+This allows Rust programmers to take advantage of type-level monads (meganads), and negative-cost abstractions (NCAs).
+
+
+
+
+# Motivation
+[motivation]: #motivation
+
+Rust currently lives in an awkward middle-ground: it's currently incredibly inexpressive, making it undesirable for the mainstream programming community which is used to languages like Idris. At the same time, Rust is also incredibly inefficient, making it undesirable for hardcore systems programmers who are used to the blazing performance of Haskell.
+
+Both of these problems can be reduced to a single issue: Rust's type system has a significant impedance mismatch with how machines work. This is why user code needs to pass through so many intermediate representations: 
+
+```
+Text -> AST -> HIR -> MIR -> LLVM-IR -> LLVM-MIR -> ASM -> Machine Code.
+```
+
+At each step the compiler is forced to try to synthesize standard machine abstractions like monads and lenses, given only sloppy imperative Rust code. This incredible journey is why introducing higher-kinded types to the language has been such a struggle: lowering the types would be dangerous at such heights. In particular, due to the Turing Completeness of Rust's type system, it's undecidable if a given lowering will reach *terminating velocity*.
+
+
+
+
+
+# Detailed design
+[design]: #detailed-design
+
+Our solution to this problem is simple: expose the compiler's pipeline to developers, so they can place types at whatever height they want, and manage the abstraction velocity themselves.
+
+Under this system, types may be made *higher* and *lower*, using the theory of *Omega Types*. The syntax for this, is simple and natural: types may be raised with the HTML `<sup>` tag, and lowered with the HTML `<sub>` tag. For example:
```

Comment:
`sup` and `sub` (and in fact, almost all HTML tags) are normal identifiers, no? This will require contextual parser that is cumbersome to implement, albeit it wouldn't be ambiguous (`Vec<sup>` not followed by ident is a type by its own).

We can instead adopt a syntax that is common enough in the running plain text and several lightweight markup languages (e.g. Pandoc's implementation of Markdown):

```
i32
Vec<i32>
Vec^i32^
Vec~i32~
Vec^Vec~i32~^
```

In the expression position, a path including lower or higher type components will be then expressed as `Vec::^i32^` etc, highly regular and unambiguous.